### ISSUE TITLE
💅 noUnnecessaryConditions false positives for optional chaining and switch cases

### ISSUE THREAD

**Issue by jakeleventhal** [biomejs/biome#10781]
### Environment information

<details>

```bash
CLI:
  Version:                      2.5.1
  Color support:                false

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_DISTRIBUTION:                npm
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     1
  TERM:                         dumb
  JS_RUNTIME_VERSION:           v24.12.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.3.14

Biome Configuration:
  Status:                       Loaded successfully.
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    suspicious/noUnnecessaryConditions

Workspace:
  Open Documents:               0
```

</details>

### Rule name

noUnnecessaryConditions

### Playground link

https://biomejs.dev/playground/?lintRules=noUnnecessaryConditions&code=dAB5AHAAZQAgAE0AbwB2AGkAZQAgAD0AIAB7AAoACQBjAGgAYQByAGEAYwB0AGUAcgBzAD8AOgAgAEEAcgByAGEAeQA8AHsACgAJAAkAaQBzAEoAZQBkAGkAOgAgAGIAbwBvAGwAZQBhAG4AOwAKAAkACQBuAGEAbQBlADoAIABzAHQAcgBpAG4AZwA7AAoACQB9AD4AOwAKAH0AOwAKAAoAZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIABvAHAAdABpAG8AbgBhAGwAQwBoAGEAaQBuAEEAcgByAGEAeQBGAGkAbAB0AGUAcgAoAG0AbwB2AGkAZQA6ACAATQBvAHYAaQBlACkAOgAgAEEAcgByAGEAeQA8AHMAdAByAGkAbgBnAD4AIAB7AAoACQBjAG8AbgBzAHQAIABqAGUAZABpAHMAIAA9ACAAbQBvAHYAaQBlAC4AYwBoAGEAcgBhAGMAdABlAHIAcwA%2FAC4AZgBpAGwAdABlAHIAKAAoAGMAaABhAHIAYQBjAHQAZQByACkAIAA9AD4AIABjAGgAYQByAGEAYwB0AGUAcgAuAGkAcwBKAGUAZABpACkAIAA%2FAD8AIABbAF0AOwAKAAoACQByAGUAdAB1AHIAbgAgAGoAZQBkAGkAcwAuAG0AYQBwACgAKABjAGgAYQByAGEAYwB0AGUAcgApACAAPQA%2BACAAYwBoAGEAcgBhAGMAdABlAHIALgBuAGEAbQBlACkAOwAKAH0ACgAKAGMAbwBuAHMAdAAgAEYAQQBDAFQASQBPAE4AUwAgAD0AIABbAAoACQB7ACAAbgBhAG0AZQA6ACAAIgBSAEUAUABVAEIATABJAEMAIgAsACAAaQBkADoAIAAxACAAfQAsAAoACQB7ACAAbgBhAG0AZQA6ACAAIgBFAE0AUABJAFIARQAiACwAIABpAGQAOgAgADIAIAB9ACwACgAJAHsAIABuAGEAbQBlADoAIAAiAEEATABMAEkAQQBOAEMARQAiACwAIABpAGQAOgAgADMAIAB9ACwACgAJAHsAIABuAGEAbQBlADoAIAAiAEMASQBTACIALAAgAGkAZAA6ACAANAAgAH0ALAAKAAkAewAgAG4AYQBtAGUAOgAgACIARgBJAFIAUwBUAF8ATwBSAEQARQBSACIALAAgAGkAZAA6ACAANQAgAH0ALAAKAAkAewAgAG4AYQBtAGUAOgAgACIAUgBFAFMASQBTAFQAQQBOAEMARQAiACwAIABpAGQAOgAgADYAIAB9ACwACgBdACAAYQBzACAAYwBvAG4AcwB0ADsACgAKAGYAdQBuAGMAdABpAG8AbgAgAGYAaQBuAGQARgBhAGMAdABpAG8AbgAoAGkAZAA6ACAAbgB1AG0AYgBlAHIAKQAgAHsACgAJAHIAZQB0AHUAcgBuACAARgBBAEMAVABJAE8ATgBTAC4AZgBpAG4AZAAoACgAZgBhAGMAdABpAG8AbgApACAAPQA%2BACAAZgBhAGMAdABpAG8AbgAuAGkAZAAgAD0APQA9ACAAaQBkACkAPwAuAG4AYQBtAGUAIAA%2FAD8AIABuAHUAbABsADsACgB9AAoACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAHMAdwBpAHQAYwBoAE8AbgBPAHAAdABpAG8AbgBhAGwAUgBlAHMAdQBsAHQAKABpAGQAOgAgAG4AdQBtAGIAZQByACkAIAB7AAoACQBzAHcAaQB0AGMAaAAgACgAZgBpAG4AZABGAGEAYwB0AGkAbwBuACgAaQBkACkAKQAgAHsACgAJAAkAYwBhAHMAZQAgACIAUgBFAFAAVQBCAEwASQBDACIAOgAKAAkACQAJAHIAZQB0AHUAcgBuACAAIgBHAGEAbABhAGMAdABpAGMAIABSAGUAcAB1AGIAbABpAGMAIgA7AAoACQAJAGMAYQBzAGUAIAAiAEUATQBQAEkAUgBFACIAOgAKAAkACQAJAHIAZQB0AHUAcgBuACAAIgBHAGEAbABhAGMAdABpAGMAIABFAG0AcABpAHIAZQAiADsACgAJAAkAZABlAGYAYQB1AGwAdAA6AAoACQAJAAkAcgBlAHQAdQByAG4AIABuAHUAbABsADsACgAJAH0ACgB9AAoACgB0AHkAcABlACAAQwBvAG4AZABpAHQAaQBvAG4AYQBsAFUAbgBpAG8AbgA8AFQAPgAgAD0AIAAiAGEAIgAgAHwAIAAiAGIAIgAgAHwAIAAoAFQAIABlAHgAdABlAG4AZABzACAAdAByAHUAZQAgAD8AIAAiAGMAIgAgADoAIABuAGUAdgBlAHIAKQA7AAoACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAGcAZQBuAGUAcgBpAGMAVQBuAGkAbwBuAFMAdwBpAHQAYwBoACgAdgBhAGwAdQBlADoAIABDAG8AbgBkAGkAdABpAG8AbgBhAGwAVQBuAGkAbwBuADwAZgBhAGwAcwBlAD4AKQAgAHsACgAJAHMAdwBpAHQAYwBoACAAKAB2AGEAbAB1AGUAKQAgAHsACgAJAAkAYwBhAHMAZQAgACIAYQAiADoACgAJAAkACQByAGUAdAB1AHIAbgAgADEAOwAKAAkACQBjAGEAcwBlACAAIgBiACIAOgAKAAkACQAJAHIAZQB0AHUAcgBuACAAMgA7AAoACQB9AAoAfQAKAA%3D%3D&language=ts

Minimal reproduction project:

```jsonc
// biome.json
{
  "$schema": "https://biomejs.dev/schemas/2.5.1/schema.json",
  "linter": {
    "enabled": true,
    "rules": {
      "preset": "none",
      "suspicious": {
        "noUnnecessaryConditions": "error"
      }
    }
  }
}
```

```ts
// src/index.ts
type Movie = {
  characters?: Array<{
    isJedi: boolean;
    name: string;
  }>;
};

export function optionalChainArrayFilter(movie: Movie): Array<string> {
  const jedis = movie.characters?.filter((character) => character.isJedi) ?? [];

  return jedis.map((character) => character.name);
}

const FACTIONS = [
  { name: "REPUBLIC", id: 1 },
  { name: "EMPIRE", id: 2 },
  { name: "ALLIANCE", id: 3 },
  { name: "CIS", id: 4 },
  { name: "FIRST_ORDER", id: 5 },
  { name: "RESISTANCE", id: 6 },
] as const;

function findFaction(id: number) {
  return FACTIONS.find((faction) => faction.id === id)?.name ?? null;
}

export function switchOnOptionalResult(id: number) {
  switch (findFaction(id)) {
    case "REPUBLIC":
      return "Galactic Republic";
    case "EMPIRE":
      return "Galactic Empire";
    default:
      return null;
  }
}

type ConditionalUnion<T> = "a" | "b" | (T extends true ? "c" : never);

export function genericUnionSwitch(value: ConditionalUnion<false>) {
  switch (value) {
    case "a":
      return 1;
    case "b":
      return 2;
  }
}
```

Steps:

```sh
bun install
bun run lint --colors=off
bunx tsc --noEmit
```

Observed diagnostics from `bun run lint --colors=off`:

```text
src/index.ts:9:16 lint/suspicious/noUnnecessaryConditions
  × Unnecessary nullish coalescing.
  i The left-hand side is guaranteed to be non-nullish, so the fallback is unreachable.

src/index.ts:29:8 lint/suspicious/noUnnecessaryConditions
  × This case is unreachable.
  i The value passed to switch can never equal this value.

src/index.ts:31:8 lint/suspicious/noUnnecessaryConditions
  × This case is unreachable.
  i The value passed to switch can never equal this value.

src/index.ts:42:8 lint/suspicious/noUnnecessaryConditions
  × This case is unreachable.
  i The value passed to switch can never equal this value.

src/index.ts:44:8 lint/suspicious/noUnnecessaryConditions
  × This case is unreachable.
  i The value passed to switch can never equal this value.
```

`bunx tsc --noEmit` passes under `strict: true`.

### Expected result

No diagnostic should be emitted for these cases.

`movie.characters` is optional, so `movie.characters?.filter(...)` can evaluate to `undefined`; the `?? []` fallback is reachable. Removing the fallback changes runtime behavior and makes `jedis.map(...)` unsafe when `characters` is absent.

`findFaction(id)` can return a faction name or `null`; `switchOnOptionalResult(1)` reaches `case "REPUBLIC"` and `switchOnOptionalResult(2)` reaches `case "EMPIRE"`, so those cases are not unreachable.

`ConditionalUnion<false>` evaluates to `"a" | "b"`; both switch cases in `genericUnionSwitch` are reachable.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct
